app/database_migrations.py

```
@@ -55,6 +99,36 @@ def _determine_schema_revision(inspector) -> str:
+    if not source.exists():
+        logger.info("Skipping startup migration backup: database file does not exist yet")
+        return None
+
+    backup_dir = source.parent / "backups"
+    backup_dir.mkdir(parents=True, exist_ok=True)
+    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
+    backup_path = backup_dir / f"{source.name}.pre_migration.{timestamp}.bak"
+
+    source_uri = f"file:{source.absolute()}?mode=ro"
+    with sqlite3.connect(source_uri, uri=True) as src_conn:
+        with sqlite3.connect(backup_path) as backup_conn:
```

Ruff (0.15.10)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)